docs(website): running docker in prod post (#2342)
glasskube/distr: website/src/content/blog/2026-04-22-running-docker-in-production.mdx

publishDate: 2026-04-22
lastUpdated: 2026-04-22
slug: 'running-docker-in-production'
authors:
  - name: 'Philip Miglinci'
    role: 'Co-Founder'
    image: '/src/assets/blog/authors/pmig.jpg'
    linkedIn: https://www.linkedin.com/in/pmigat/
    gitHub: https://github.com/pmig
image: '/src/assets/blog/2026-04-22-running-docker-in-production/thumb.png'
tags:
  - Docker
  - Docker Compose
  - Docker Swarm
  - Production
  - DevOps
  - Security
---

import {Aside} from '@astrojs/starlight/components';
import EmbeddedWhitePaperCta from '~/components/cta/embedded/EmbeddedWhitePaperCta.astro';

I am Philip—an engineer working at Distr, which helps software and AI companies distribute their applications to self-managed environments.
Our Open Source Software Distribution platform is available on GitHub ([`github.com/distr-sh/distr`](https://github.com/distr-sh/distr)) and orchestrates both Docker Compose and Docker Swarm deployments on customer hosts every day.

> Most of the production incidents I have seen on Docker Compose hosts come from the same handful of quirks: an old container that should have been removed, a disk that filled up overnight, a health check that detected a problem and then did nothing about it, a `:latest` tag that pointed somewhere new, or a socket mount nobody thought twice about. None of these are bugs in Docker. They are deliberate trade-offs in a tool that started as a developer experience and is now running the back end of a lot of real businesses. This post collects the recurring ones, with the commands and the operational answer for each.

Reviewer: > a tool that started as a developer experience

doesn't make much sense to me. how is a tool an experience?

<hr />

## Why Docker Compose Still Runs a Lot of Production

> Before the list of quirks, a quick word on the audience. The sweet spot for Docker Compose in production is the single-node deployment that ships as a self-contained artifact: a vendor pushing a multi-container application into a customer environment, an internal team running a long-tail service that does not justify a Kubernetes cluster, an edge box in a retail location. The footprint is small, the operational overhead is low, and a competent operator can reason about the whole stack from one `docker-compose.yaml`.

Reviewer: > self-contained artifact

compose is not self-contained

> That posture is exactly why the quirks bite. Compose does not have a control plane reminding you to clean up images or restart unhealthy containers, and it does not assume you have one. It assumes you—or whoever is deploying it—will. If you ship Compose files to customers, the safe assumption is that the customer will not. The rest of this post is about closing that gap, either by hand or with an [agent](/docs/agents/docker-agent) that does it for you. If you have already concluded that the gap is too wide and you want a comparison with the next step up, [Docker Compose vs Kubernetes](/blog/docker-compose-vs-kubernetes/) is the deeper read.

Reviewer: > That posture is exactly why the quirks bite.

Maybe this is just my English not being good enough but I'm not sure what posture means here

> That posture is exactly why the quirks bite. Compose does not have a control plane reminding you to clean up images or restart unhealthy containers, and it does not assume you have one. It assumes you—or whoever is deploying it—will. If you ship Compose files to customers, the safe assumption is that the customer will not. The rest of this post is about closing that gap, either by hand or with an [agent](/docs/agents/docker-agent) that does it for you. If you have already concluded that the gap is too wide and you want a comparison with the next step up, [Docker Compose vs Kubernetes](/blog/docker-compose-vs-kubernetes/) is the deeper read.

Reviewer: > If you ship Compose files to customers, the safe assumption is that the customer will not. 

will not what? also ship compose files?

> That posture is exactly why the quirks bite. Compose does not have a control plane reminding you to clean up images or restart unhealthy containers, and it does not assume you have one. It assumes you—or whoever is deploying it—will. If you ship Compose files to customers, the safe assumption is that the customer will not. The rest of this post is about closing that gap, either by hand or with an [agent](/docs/agents/docker-agent) that does it for you. If you have already concluded that the gap is too wide and you want a comparison with the next step up, [Docker Compose vs Kubernetes](/blog/docker-compose-vs-kubernetes/) is the deeper read.

Reviewer: > The rest of this post is about closing that gap

What gap? maybe missing in the sentence before

Remove a service from `docker-compose.yaml`, run `docker compose up -d`, and the container you removed keeps running. It is detached from the project but still bound to the same networks and ports. `docker compose ps` will not show it, because Compose only lists what is in the current file. `docker ps --filter label=com.docker.compose.project=<name>` will, because Docker still has the label on the container. This is how you discover, six months in, that an old `worker` service has been quietly consuming RAM since the last refactor.

The fix is one flag:

```bash
docker compose up -d --remove-orphans
docker compose down --remove-orphans
```

The flag tells Compose: anything that was once part of this project but is no longer in the file should be removed. To audit before deleting, list everything Docker still associates with the project name:

```bash
docker ps -a --filter label=com.docker.compose.project=<name>
```

<Aside type="note">
  `--remove-orphans` only touches containers that share the Compose project
  label. It will not remove unrelated containers on the host. The risk is
  bounded to containers Compose itself created earlier under this project name.
</Aside>

> In Distr's Docker agent we pass `RemoveOrphans: true` on every Compose `Up` call, so customer hosts never accumulate orphans across deployment updates. That single flag has eliminated a recurring class of "the old version is still answering on port 8080" support tickets.

Reviewer: ```suggestion
Distr's Docker agent passes `RemoveOrphans: true` on every Compose `Up` call, so customer hosts never accumulate orphans across deployment updates. That single flag has eliminated a recurring class of "the old version is still answering on port 8080" support tickets.
```

docker builder prune -f                           # drop the BuildKit cache
```

`docker volume prune -f` exists too, and it is genuinely useful, but read the next aside before you run it.

<Aside type="caution">
  `docker volume prune` removes **every volume not currently attached to a
  container**, including named volumes you created on purpose. If you stop a
  database container to do maintenance and run `docker volume prune` while it is
  down, you can lose its data. Use `docker volume ls --filter dangling=true`
  first to see what is actually unreferenced, and prefer named-volume management
  over blanket prunes on production hosts.
</Aside>

The other half of the disk story is logs. Cap them at the daemon level, once, in `/etc/docker/daemon.json`:

```json
{
  "log-driver": "json-file",
  "log-opts": {
    "max-size": "10m",
    "max-file": "3"
  }
}
```

After `systemctl restart docker`, every new container will rotate its logs at 10 MB and keep at most three rotated files—30 MB ceiling per container, instead of "until the disk is gone." Existing containers need to be recreated to pick up the new defaults.

> This is one of the topics worth getting right before you ship. If you also need to ship config files alongside the Compose file (a `daemon.json` is one such file, but most apps have their own), the patterns in [Mounting Configuration Files in Docker Compose](/blog/docker-compose-mount-config-files/) cover the four approaches I have seen survive customer deployments.

Reviewer: The linked article here doesn't really work for the docker daemon config

2. **Run on Docker Swarm.** Swarm restarts unhealthy tasks by default. If you are already considering Swarm, this is one of the better reasons.
3. **Use Distr.** Every Distr Docker agent deploys an adapted autoheal service alongside it. The "Enable autoheal for all containers" toggle is on by default at deployment-target creation, so customer-side restarts of unhealthy containers happen without anyone configuring it.

<Aside type="note">
  In Distr the autoheal setting is fixed at deployment-target creation and
  cannot be toggled later. If you need to change it on an existing target,
  recreate the target. Worth knowing before you provision.
</Aside>

Whichever path you pick, the takeaway is the same: a `HEALTHCHECK` without something acting on it is a status light, not a self-healing system.

<EmbeddedWhitePaperCta />

## `:latest` Is a Lie — Pin by Digest

> Docker tags are mutable references. `myapp:1.4` today is whatever the registry currently has under that tag; tomorrow it can point at a different layer set after a re-push. `:latest` is the worst offender because everyone treats it as a synonym for "stable" when in practice it means "whatever was pushed most recently." The result, in production, is that two hosts pulling the "same" tag five minutes apart can end up running different code.

Reviewer: ```suggestion
Docker tags are mutable references. `myapp:1.4` today is whatever the registry currently has under that tag; tomorrow it can point at a different layer set after a re-push. `:latest` is the worst offender because everyone treats it as a synonym for "stable" when in practice it often means "whatever was pushed most recently." The result, in production, is that two hosts pulling the "same" tag five minutes apart can end up running different code.
```

The fix is to pin by content-addressable digest. Every image has one, and Docker accepts it anywhere a tag would go.

> To find the digest for an image you already pulled (works only for images that have been pulled from or pushed to a registry):

Reviewer: ```suggestion
To find the digest for an image you already pulled:
```

The first part is redundant and the second part is wrong

```bash
docker image inspect --format='{{index .RepoDigests 0}}' myapp:1.4
# myapp@sha256:9b7c…
```

Or, without pulling, from the local Docker installation against the remote registry:

```bash
docker buildx imagetools inspect myapp:1.4
```

In your Compose file, replace the tag with the digest:

```yaml
services:
  app:
    image: myapp@sha256:9b7c0a3e1f...
```

A pull against a digest fails fast if the registry no longer has those bytes, which is exactly what you want—silent drift becomes a loud error. The same image reference works in `docker stack deploy`, in `docker run`, and in Kubernetes manifests.

For the broader picture of what your customers can extract from a published image (and why image hygiene matters beyond reproducibility), [How to Protect Your Source Code and IP in Docker and Kubernetes Deployments](/blog/protect-source-code-docker-container/) is the companion read. And if the registry choice itself is up in the air, the [container registry comparison](/blog/container-image-registry-comparison/) covers the trade-offs.

## Mounting the Docker Socket Is Root Equivalence

> A container with `/var/run/docker.sock` mounted can call the Docker API, and the Docker API can launch a privileged container that mounts the host's root filesystem. In other words: any container with the socket has root on the host. This is not a Docker bug; it is the threat model of the socket. It deserves a moment of attention because the line that grants this access is one bind mount in a Compose file and is easy to add without thinking about it.

Reviewer: ```suggestion
A container with `/var/run/docker.sock` mounted can call the Docker API, and the Docker API can launch a privileged container that mounts the host's root filesystem. In other words: any container with the socket has effectively root privileges on the host. This is not a Docker bug; it is the threat model of the socket. It deserves a moment of attention because the line that grants this access is one bind mount in a Compose file and is easy to add without thinking about it.
```

<Aside type="caution">
  Mounting `/var/run/docker.sock` into a container is equivalent to giving that

>   container root on the host. A `:ro` flag does **not** help—the socket is

Reviewer: ```suggestion
  container root privileges on the host. A `:ro` flag does **not** help—the socket is
```

  bidirectional RPC, not a file you read. Treat every container that mounts the

>   socket as a tier-one trust boundary.

Reviewer: is everyone expected to know what a tier-one trust boundary is? I don't 👀

</Aside>

Practical hygiene:

- **Inventory the containers that mount the socket.** Agents, CI runners, monitoring sidecars, container management UIs—keep the list short and intentional.
- **Run rootless Docker where possible.** [`dockerd-rootless-setuptool.sh install`](https://docs.docker.com/engine/security/rootless/) sets up a Docker daemon that runs as a regular user. The blast radius of a compromised socket-mounting container shrinks from "full host" to "this user account."
- **Consider socket-proxy.** Projects like Tecnativa's `docker-socket-proxy` expose a filtered subset of the API to the container that needs it (e.g. read-only `containers` and `events` for monitoring) instead of the full socket.
- **Keep socket-mounting images minimal.** Smaller surface, fewer libraries, fewer ways in.

> The Distr Docker agent does mount the socket—it has to, in order to orchestrate Compose and Swarm on the host. We document that boundary openly in the [Docker agent docs](/docs/agents/docker-agent) so customer security teams can review it before installation. The agent authenticates to the Hub with a JWT, the install secret is shown once and never stored, and the host's `~/.docker` directory is mounted read-only so existing registry credentials are reused without being copied.

Reviewer: > the host's `~/.docker` directory is mounted read-only so existing registry credentials are reused without being copied

this doesn't make sense to me... given `ro` access, what prevents me from making a copy?

## There Is No Remote-Update Button in Docker

> `docker compose pull && docker compose up -d` is a fine command if you are SSH'd into the host. At customer scale—dozens of self-managed environments behind firewalls, each with its own change-control process—it is the wrong shape of solution. Docker has no built-in mechanism to push a new manifest to a running host from somewhere else. Docker Hub webhooks can trigger a CI rebuild when an image is pushed, but they do not reach into a customer's network and tell their `docker compose` to pull.

Reviewer: > shape of solution

sounds weird

The usual workarounds and what they cost:

- **[Watchtower](https://github.com/containrrr/watchtower):** Polls the registry on a schedule, pulls new images, recreates containers. Easy to set up, hard to control. No staged rollout, no rollback path, limited visibility from your side—you find out a customer updated when they file a ticket.
- **Bastion + SSH + Ansible/scripts:** Works for ten customers. Falls apart at fifty, especially when three of them are air-gapped and four run their own change-control cadence. Every operator has to live with shared keys and a maintenance window calendar.

> - **A pull-based agent.** This is the shape Distr lands on. The agent runs on the customer host, polls a known endpoint at a configurable interval (5 seconds by default), and reconciles the local Compose state against what the Hub says it should be. The agent reports status back, so you can see in your dashboard which customers are on which version. When the agent itself needs to update, it spawns a separate container to perform the swap so it is not trying to replace itself while running.

Reviewer: > polls a known endpoint at a configurable interval (5 seconds by default)

it's not configurable I think

The pattern is not unique—Kubernetes operators and GitOps tools do the same thing—but Compose users routinely re-invent it badly. If you find yourself building one, at least give it rollback, status reporting, and a way to pin versions, or you will end up with a fleet that drifts in ways you cannot see.

> The other thing worth noting: scheduled jobs run _inside_ the application also have no native Compose answer. If your update story includes "run this migration once after the deploy," you will eventually need to pick one of [the three approaches I cover for Compose cron jobs](/blog/docker-compose-cron-jobs/).

Reviewer: > scheduled jobs run _inside_ the application also have no native Compose answer

?

runnking jobs inside the application seems like the answer?

> The other thing worth noting: scheduled jobs run _inside_ the application also have no native Compose answer. If your update story includes "run this migration once after the deploy," you will eventually need to pick one of [the three approaches I cover for Compose cron jobs](/blog/docker-compose-cron-jobs/).

Reviewer: > If your update story includes "run this migration once after the deploy," you will eventually need to pick one of the three approaches I cover for Compose cron jobs

this is a bad example for a cronjob

  Swarm is not an industry standard. Adoption has plateaued, third-party tooling
  is thin, and most cloud providers have moved their managed offerings to
  Kubernetes. It can still be the right call for a self-contained on-prem
  cluster, but pick it eyes-open—for most teams outgrowing Compose, Kubernetes
  is the safer long-term bet.
</Aside>

The Distr agent supports both — the Hub records whether a deployment is Compose or Swarm, and the agent runs the matching `docker compose up` or `docker stack deploy`. If you do choose Swarm, the routing-and-Traefik write-up is in [Load Balancing and Routing in Docker Swarm](/blog/docker-swarm-load-balancing-routing/), and the product walkthrough is in [Distributing Applications to Docker Swarm](/blog/distr-docker-swarm/).

## Docker in Production Is a Solved Problem If You Know the Shape of the Problem

None of the quirks above are secret. They are all in Docker's documentation, in GitHub issues that have been open for years, and in the war stories of every team that has run Compose in anger. What makes them dangerous is not the quirks themselves but the order in which you discover them: usually at 2 a.m., one at a time.

> The summary version, in case you want to leave a note for the next person:

Reviewer: this sentence also sounds weird IMO
